client/optimizing.go

```diff
@@ -330,88 +331,200 @@ type watchResult struct {
+		stat := requestStat{
+			client:    r.Client,
+			rtt:       time.Since(timeOfRound),
+			startTime: timeOfRound,
 		}
-		if ctx.Err() != nil {
-			done.Done()
-			return
+		oc.updateStats([]*requestStat{&stat})
+		if round > latest {
+			latest = round
+			out <- r.Result
 		}
+	}
+}
 
-		oc.log.Warn("optimizing_client", "watch channel closed")
+// Watch returns new randomness as it becomes available.
+func (oc *optimizingClient) Watch(ctx context.Context) <-chan Result {
+	outChan := make(chan Result, defaultChannelBuffer)
+	inChan := make(chan watchResult, defaultChannelBuffer)
 
-		if oc.watchRetryInterval < 0 { // negative interval disables retries
-			done.Done()
-			return
-		}
+	info, err := oc.Info(ctx)
+	if err != nil {
+		oc.log.Error("optimizing_client", "failed to learn info", "err", err)
+		close(outChan)
+		return outChan
+	}
 
-		t := time.NewTimer(oc.watchRetryInterval)
-		select {
-		case <-t.C:
-		case <-ctx.Done():
-			t.Stop()
-		}
+	state := watchState{
+		ctx:           ctx,
+		optimizer:     oc,
+		active:        make([]watchingClient, 0),
+		failed:        make([]failedClient, 0),
+		retryInterval: oc.watchRetryInterval,
 	}
+
+	go state.loop(inChan)
+	go oc.trackWatchResults(info, inChan, outChan)
+	return outChan
 }
 
-func (oc *optimizingClient) trackWatchResults(info *chain.Info, in chan watchResult, out chan Result, done chan bool) {
-	defer close(out)
+type watchingClient struct {
+	Client
+	context.CancelFunc
+}
 
-	latest := uint64(0)
+type failedClient struct {
+	Client
+	time.Time
+}
+
+type watchState struct {
+	ctx           context.Context
+	optimizer     *optimizingClient
+	active        []watchingClient
+	failed        []failedClient
+	retryInterval time.Duration
+}
+
+func (ws *watchState) loop(resultChan chan watchResult) {
+	closingClients := make(chan Client, 1)
+	defer close(resultChan)
+
+	// spin up initial watcher(s)
+	ws.tryRepopulate(resultChan, closingClients)
+
+	ticker := time.NewTicker(ws.optimizer.watchRetryInterval)
+	defer ticker.Stop()
 	for {
 		select {
-		case r, ok := <-in:
-			if !ok {
+		case c := <-closingClients:
+			// replace failed watchers
+			ws.done(c)
+			if ws.ctx.Err() == nil {
+				ws.tryRepopulate(resultChan, closingClients)
+			}
+			if len(ws.active) == 0 {
 				return
 			}
-
-			round := r.Result.Round()
-			timeOfRound := time.Unix(chain.TimeOfRound(info.Period, info.GenesisTime, round), 0)
-			stat := requestStat{
-				client:    r.Client,
-				rtt:       time.Since(timeOfRound),
-				startTime: timeOfRound,
+		case <-ticker.C:
+			// periodically cycle to fastest client.
+			if ws.has(ws.optimizer.fastestClients()[0]) == -1 {
+				ws.closeSlowest()
+				ws.tryRepopulate(resultChan, closingClients)
 			}
-			oc.updateStats([]*requestStat{&stat})
-			if round > latest {
-				latest = round
-				out <- r.Result
+		case <-ws.ctx.Done():
+			for _, c := range ws.active {
+				c.CancelFunc()
```

Comment:
Do we need to worry about canceling these and them all then trying to write to `closingClients` which we'll not read?

Reply:
why won't we read it? added a comment to clarify

Reply:
`ws.active` is set to an empty slice of watchingClient immediately below

Reply:
good catch!